core/src/smartcontracts/isi/asset.rs

```diff
@@ -54,89 +54,165 @@ pub mod isi {
+            Ok(self.checked_sub(rhs)?)
+        }
+    }
+
+    /// Trait for collecting fields associated with asset isi instructions:
+    /// - asset value type
+    /// - default asset value
+    trait AssetInstructionInfo {
+        const EXPECTED_VALUE_TYPE: AssetValueType;
+        const DEFAULT_ASSET_VALUE: Self;
+    }
+
+    macro_rules! impl_asset_instruction_info {
+        ($ty:ty, $expected_value_type:expr, $default_value:expr) => {
+            impl AssetInstructionInfo for $ty {
+                const EXPECTED_VALUE_TYPE: AssetValueType = $expected_value_type;
+                const DEFAULT_ASSET_VALUE: Self = $default_value;
+            }
+        };
+    }
+
+    impl_asset_instruction_info!(u32, AssetValueType::Quantity, 0_u32);
+    impl_asset_instruction_info!(u128, AssetValueType::BigQuantity, 0_u128);
+    impl_asset_instruction_info!(Fixed, AssetValueType::Fixed, Fixed::ZERO);
+
+    /// Trait for blanket mint implementation.
+    trait InnerMint {
+        fn execute<W: WorldTrait, Err>(
+            mint: Mint<Asset, Self>,
             _authority: <Account as Identifiable>::Id,
             wsv: &WorldStateView<W>,
-        ) -> Result<(), Self::Error> {
-            let asset_id = self.destination_id;
-
-            assert_can_mint(&asset_id.definition_id, wsv, AssetValueType::Fixed)?;
-            wsv.asset_or_insert(&asset_id, Fixed::ZERO)?;
+        ) -> Result<(), Err>
+        where
+            Self: AssetInstructionInfo + TryCheckedAdd + FloatMetricFrom + Copy,
+            AssetValue: From<Self> + TryAsMut<Self>,
+            Value: From<Self>,
+            <AssetValue as TryAsMut<Self>>::Error: std::error::Error + Send + Sync + 'static,
+            Err: From<Error>,
+        {
+            let asset_id = mint.destination_id;
+
+            assert_can_mint(
+                &asset_id.definition_id,
+                wsv,
+                <Self as AssetInstructionInfo>::EXPECTED_VALUE_TYPE,
+            )?;
+            wsv.asset_or_insert(
+                &asset_id,
+                <Self as AssetInstructionInfo>::DEFAULT_ASSET_VALUE,
+            )?;
             wsv.modify_asset(&asset_id, |asset| {
-                let quantity: &mut Fixed = asset
+                let quantity: &mut Self = asset
                     .try_as_mut()
                     .map_err(eyre::Error::from)
                     .map_err(|e| Error::Conversion(e.to_string()))?;
-                *quantity = quantity.checked_add(self.object)?;
-                wsv.metrics.tx_amounts.observe((*quantity).into());
+                *quantity = quantity.try_checked_add(mint.object)?;
+                wsv.metrics
+                    .tx_amounts
+                    .observe(<Self as FloatMetricFrom>::from(*quantity));
 
                 Ok(AssetEvent::Added(asset_id.clone()))
-            })
+            })?;
+            Ok(())
         }
     }
 
+    macro_rules! impl_mint {
+        ($ty:ty, $error:ty, $metrics:literal) => {
```

Comment:
```suggestion
        ($ty:ty, $metrics:literal) => {
```
Below you hard-code Error. The macro invocation repeats less.